sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -119,9 +122,27 @@ object ProfilerOnExecutor extends Logging {
       val stageId = taskCtx.stageId
       if (stageRanges.contains(stageId)) {
         synchronized {
-          activeStages.add(taskCtx.stageId)
-          enable()
-          startPollingDriver()
+          if (taskLimit <= 0) {
+            // Unlimited tasks per stage
+            activeStages.add(taskCtx.stageId)
+            enable()
+            startPollingDriver()
+          } else {
+            // Limited tasks per stage
+            if (stageTaskCount.getOrElse(stageId, 0) < taskLimit) {
+              activeStages.add(taskCtx.stageId)
+              enable()
+              startPollingDriver()
+            }
+            taskCtx.addTaskCompletionListener[Unit] { _ =>
+              val currentCount = stageTaskCount.getOrElse(stageId, 0)
+              if (currentCount < taskLimit) {
+                stageTaskCount(stageId) = currentCount + 1
```

Comment:
stageId is leaked in the map and never removed.  Can solve that by storing a 0 in the map for the stage when the first task for a stage to be profiled is seen.  Task completion callback would try to lookup stage in the map.  If it gets nothing, there's nothing for it to do.  If it gets a task count, it increments it.  If we're still under the limit then it stores the updated count, otherwise it removes the count for that stage from the map.

Reply:
I have not found a good way to check if a task is the first task seen for a stage, so for now I clean the `stageTaskCount` in `updateActiveFromDriver`.